packages/core/src/credential/Credential.ts

```diff
@@ -81,103 +277,158 @@ export function verifyAgainstCType(
-    request,
-    attestation,
+    claim,
+    legitimations,
+    claimHashes,
+    claimNonceMap,
+    rootHash,
+    delegationId,
   }
   verifyDataStructure(credential)
   return credential
 }
 
-/**
- * Custom Type Guard to determine input being of type ICredential.
- *
- * @param input The potentially only partial ICredential.
- *
- * @returns Boolean whether input is of type ICredential.
- */
-export function isICredential(input: unknown): input is ICredential {
-  try {
-    verifyDataStructure(input as ICredential)
-  } catch (error) {
-    return false
-  }
-  return true
+type VerifyOptions = {
+  ctype?: ICType
+  challenge?: string
+  resolver?: IDidResolver
+  allowUnsigned?: boolean
 }
 
 /**
- * Verifies whether the credential is valid. It is valid if:
- * * the data is valid (see [[verifyDataIntegrity]]);
- * and
- * * the [[Attestation]] object for this credential is valid (see [[Attestation.checkValidity]], where the **chain** is queried).
+ * Verifies data structure, data integrity and claimers signature.
  *
  * Upon presentation of a credential, a verifier would call this [[verify]] function.
  *
- * @param credential - The credential to check for validity.
- * @param verificationOpts The additional options to use upon attested credential verification.
- * @param verificationOpts.resolver - The resolver used to resolve the claimer's identity if it is not passed in.
- * Defaults to [[DidResolver]].
- * @param verificationOpts.challenge - The expected value of the challenge. Verification will fail in case of a mismatch.
- * @returns A promise containing whether the provided credential is valid.
+ * @param credential - The object to check.
+ * @param options - Additional parameter for more verification steps.
+ * @param options.ctype - CType which the included claim should be checked against.
+ * @param options.challenge -  The expected value of the challenge. Verification will fail in case of a mismatch.
+ * @param options.allowUnsigned - Wether to check the signature without a challenge being present.
+ * @param options.resolver - The resolver used to resolve the claimer's identity. Defaults to [[DidResolver]].
  */
 export async function verify(
   credential: ICredential,
-  verificationOpts: {
-    resolver?: IDidResolver
-    challenge?: string
-  } = {}
-): Promise<boolean> {
-  return (
-    verifyDataIntegrity(credential) &&
-    (await RequestForAttestation.verifySignature(
-      credential.request,
-      verificationOpts
-    )) &&
-    Attestation.checkValidity(credential.attestation)
-  )
+  {
+    ctype,
+    challenge,
+    resolver = DidResolver,
+    allowUnsigned = false,
+  }: VerifyOptions = {}
+): Promise<void> {
+  verifyDataStructure(credential)
+  verifyDataIntegrity(credential)
+
+  if (ctype) {
+    const isSchemaValid = verifyAgainstCType(credential, ctype)
+    if (!isSchemaValid) throw new SDKErrors.CredentialUnverifiableError()
+  }
+
+  if (challenge || !allowUnsigned) {
```

Comment:
why do we check for the challenge though?

Reply:
There might be no challenge and no signature. In that case a user might still not want it to fail on verification, e.g. when using credentials in did. service endpoints

Reply:
I don't understand, if I set allowUnsigned to false then it MUST fail if unsigned, right?

Reply:
```
allowUnsigned = true -> signed -> pass
allowUnsigned = true -> unsigned -> pass
allowUnsigned = false -> signed -> pass
allowUnsigned = false -> unsigned -> fail
```
right?

Reply:
yes, that what it should do

Reply:
I get that you want to force allowUnsigned = true if the user passes a challenge - makes sense. So here's what we need:
1. Check signature ALWAYS if there is one
2. fail if allowUnsigned = false and no signature
3. fail if challenge and no signature

Problem is, you don't check the signature at all if allowUnsigned = true and no challenge. Instead, make an extra branch that simply throws for cases 2 & 3, then always check the signature if present.

Reply:
https://github.com/KILTprotocol/sdk-js/pull/587